plugins/consensus/opiniongiver.go

```diff
@@ -0,0 +1,168 @@
+package consensus
+
+import (
+	"context"
+	"fmt"
+	"net"
+	"strconv"
+	"time"
+
+	"github.com/golang/protobuf/proto"
+	"github.com/iotaledger/goshimmer/dapps/valuetransfers"
+	"github.com/iotaledger/goshimmer/dapps/valuetransfers/packages/branchmanager"
+	"github.com/iotaledger/goshimmer/packages/metrics"
+	"github.com/iotaledger/goshimmer/packages/vote"
+	votenet "github.com/iotaledger/goshimmer/packages/vote/net"
+	"github.com/iotaledger/goshimmer/packages/vote/statement"
+	"github.com/iotaledger/goshimmer/plugins/autopeering"
+	"github.com/iotaledger/hive.go/autopeering/peer"
+	"github.com/iotaledger/hive.go/autopeering/peer/service"
+	"google.golang.org/grpc"
+)
+
+// OpinionGiver is a wrapper for both statements and peers.
+type OpinionGiver struct {
+	id   string
+	view *statement.View
+	pog  *PeerOpinionGiver
+}
+
+// OpinionGivers is a map of OpinionGiver.
+type OpinionGivers map[string]OpinionGiver
+
+// Query retrievs the opinions about the given conflicts and timestamps.
+func (o *OpinionGiver) Query(ctx context.Context, conflictIDs []string, timestampIDs []string) (opinions vote.Opinions, err error) {
+	for i := 0; i < waitForStatement; i++ {
+		opinions, err = o.view.Query(ctx, conflictIDs, timestampIDs)
+		if err == nil {
+			return opinions, nil
+		}
+		time.Sleep(time.Second)
+	}
+
+	opinions, err = o.pog.Query(ctx, conflictIDs, timestampIDs)
+	if err != nil {
+		return nil, err
+	}
+
+	return opinions, nil
+}
+
+// ID returns a string representation of the identifier of the underlying Peer.
+func (o *OpinionGiver) ID() string {
+	return o.id
+}
+
+// OpinionGiverFunc returns a slice of opinion givers.
+func OpinionGiverFunc() (givers []vote.OpinionGiver, err error) {
+	opinionGiversMap := make(map[string]*OpinionGiver)
+	opinionGivers := make([]vote.OpinionGiver, 0)
+
+	for _, v := range Registry().NodesView() {
+		opinionGiversMap[v.ID()] = &OpinionGiver{
+			id:   v.ID(),
+			view: v,
+		}
+	}
+
+	for _, p := range autopeering.Discovery().GetVerifiedPeers() {
+		fpcService := p.Services().Get(service.FPCKey)
+		if fpcService == nil {
+			continue
+		}
+		if _, ok := opinionGiversMap[p.ID().String()]; !ok {
+			opinionGiversMap[p.ID().String()] = &OpinionGiver{
+				id: p.ID().String(),
+			}
+		}
+		opinionGiversMap[p.ID().String()].pog = &PeerOpinionGiver{p: p}
+	}
+
+	for _, v := range opinionGiversMap {
```

Comment:
I assume you copy here for thread safety. Please add a comment. Also for each `for` loop above a short comment would be nice.